master: osc/ucx fix dynamic windows

janjust opened the pull request:
Signed-off-by: Tomislav Janjusic <tomislavj@nvidia.com>

abouteiller committed:
Add a devel document describing requirements for FT in Open MPI components

Update to the devel document for adding FT support to components. This
commit adds information about wait-sync, requests, BTLs, Files, RMA

Signed-off-by: Aurélien Bouteiller <bouteill@icl.utk.edu>

Complete the description of how to test for FT and how to update
OFI and UCX PML/MTL components

add figure to the devel FT doc

Move documentation in a subdir of contrib

Update the URL for ulfm-testings

Signed-off-by: Aurélien Bouteiller <bouteill@icl.utk.edu>

Artemy-Mellanox committed:
pml/ucx: pml_ucx_multi_send_nb support

Enable passing multi-send optimization flag for nonblocking operations.

Signed-off-by: Artemy Kovalyov <artemyko@mellanox.com>

Commit:
Adding new oshmem APIs for oshmem v1.5

Signed-off-by:  Mamzi Bayatpour                 <mbayatpour@nvidia.com>
Co-authored-by: Tomislav Janjusic               <tomislavj@nvidia.com>
Co-authored-by: Manjunath Gorentla Venkata      <manjunath@nvidia.com>

janjust committed:
osc/ucx fixes dynamic windows, 1byte ops, outstanding inflight ops resulting in assertion failures

Co-authored-by: Artem Polyakov <artpol84@gmail.com>

Signed-off-by: Tomislav Janjusic <tomislavj@nvidia.com>

yosefe committed:
Merge pull request #9610 from Artemy-Mellanox/topic/multi_send

pml/ucx: pml_ucx_multi_send_nb support

janjust commented on ompi/mca/osc/ucx/osc_ucx_passive_target.c:
```diff
@@ -216,20 +216,33 @@ int ompi_osc_ucx_lock_all(int mpi_assert, struct ompi_win_t *win) {
+        }
     }
 
+
```
I'll remove this useless whitespace

janjust replied:
```suggestion
```

janjust commented on opal/mca/common/ucx/common_ucx_wpool.c:
```diff
@@ -447,12 +448,14 @@ int opal_common_ucx_wpmem_create(opal_common_ucx_ctx_t *ctx, void **mem_base, si
+        }
     }
-
+    
```
same here - I will remove them, if you see more, please let me know

janjust replied:
```suggestion
```

janjust commented on opal/mca/common/ucx/common_ucx_wpool.c:
```diff
@@ -634,11 +637,16 @@ static int _tlocal_ctx_connect(_ctx_record_t *ctx_rec, int target)
         opal_mutex_unlock(&winfo->mutex);
         return OPAL_ERROR;
     }
+
+    /* Finish ep setup */
+    while(progress) {
+        progress = ucp_worker_progress(winfo->worker);
+    }
+
```
wops - made a mistake here - this is not necessary, I'll remove it

janjust replied:
```suggestion
```

janjust commented on opal/mca/common/ucx/common_ucx_wpool.h:
```diff
@@ -298,9 +304,6 @@ static inline int _periodical_flush_nb(opal_common_ucx_wpmem_t *mem, opal_common
     int rc = OPAL_SUCCESS;
 
-    winfo->inflight_ops[target]++;
-    winfo->global_inflight_ops++;
```
in case you are wondering, why this was removed - I moved them outside - to the calls which actually need to increase inflight ops, if they are in here, I noticed that it will increase inflights ops after an win_unlock, and it will cause an assertion failure

artpol84 replied:
let's discuss tomorrow. Maybe better having a flag to prevent them to be increased here, still better than remembering to increment them outside

janjust replied:
We can discuss it, but, the flush call should not increment operations, imo. It's a flush, and all it needs to do is to flush when the condition is met. If you post an operation, you should increment inflight ops in case they are outstanding.

artpol84 commented on ompi/mca/osc/ucx/osc_ucx_passive_target.c:
```diff
@@ -216,20 +216,33 @@ int ompi_osc_ucx_lock_all(int mpi_assert, struct ompi_win_t *win) {
-    int ret = OMPI_SUCCESS;
+    int comm_size = ompi_comm_size(module->comm),
+        i = 0,
+        ret = OMPI_SUCCESS;
```
```suggestion
    int i = 0, ret = OMPI_SUCCESS;
```

artpol84 commented on ompi/mca/osc/ucx/osc_ucx_passive_target.c:
```diff
@@ -216,20 +216,33 @@ int ompi_osc_ucx_lock_all(int mpi_assert, struct ompi_win_t *win) {
-    ret = opal_common_ucx_wpmem_flush(module->mem, OPAL_COMMON_UCX_SCOPE_WORKER, 0);
-    if (ret != OMPI_SUCCESS) {
-        return ret;
+    if (module->flavor == MPI_WIN_FLAVOR_DYNAMIC) {
```
Good catch

artpol84 commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -353,20 +353,20 @@ static inline int get_dynamic_win_info(uint64_t remote_addr, ompi_osc_ucx_module
+        module->local_dynamic_win_info[*dynamic_win_index].mem->mem_addrs = calloc(ompi_comm_size(module->comm),
                                                                         OMPI_OSC_UCX_MEM_ADDR_MAX_LEN);
-        module->local_dynamic_win_info[contain].mem->mem_displs =calloc(ompi_comm_size(module->comm),
+        module->local_dynamic_win_info[*dynamic_win_index].mem->mem_displs =calloc(ompi_comm_size(module->comm),
```
```suggestion
        module->local_dynamic_win_info[*dynamic_win_index].mem->mem_displs = calloc(ompi_comm_size(module->comm),
```

artpol84 commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -325,18 +325,18 @@ static inline int end_atomicity(
 
 static inline int get_dynamic_win_info(uint64_t remote_addr, ompi_osc_ucx_module_t *module,
-                                       int target) {
+                                       int target, int *dynamic_win_index) {
```
This is a minor suggestion and can be ignored, but having a short name usually improves the readability IMO.
The function name assumes getting "dynamic_win_info" so no need to preserve it in the argument (again IMO)
```suggestion
                                       int target, int *win_idx) {
```

artpol84 commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -325,18 +325,18 @@ static inline int end_atomicity(
     uint64_t remote_state_addr = (module->state_addrs)[target] + OSC_UCX_STATE_DYNAMIC_WIN_CNT_OFFSET;
-    size_t len = sizeof(uint64_t) + sizeof(ompi_osc_dynamic_win_info_t) * OMPI_OSC_UCX_ATTACH_MAX;
-    char *temp_buf = malloc(len);
+    size_t remote_state_addr_len = sizeof(uint64_t) + sizeof(ompi_osc_dynamic_win_info_t) * OMPI_OSC_UCX_ATTACH_MAX;
```
Also IMO and can be ignored.
BTW, this is not jsut address, its more a state, right?
```suggestion
    size_t rem_state_len = sizeof(uint64_t) + sizeof(ompi_osc_dynamic_win_info_t) * OMPI_OSC_UCX_ATTACH_MAX;
```

artpol84 replied:
but in general, it's just the var name change and not more than that

janjust replied:
you are correct, not just the length of the address

artpol84 commented on opal/mca/common/ucx/common_ucx_wpool.c:
```diff
@@ -447,12 +448,14 @@ int opal_common_ucx_wpmem_create(opal_common_ucx_ctx_t *ctx, void **mem_base, si
-                        exchange_metadata);
-    if (ret != OPAL_SUCCESS) {
-        goto error_rkey_pack;
+    if (exchange_mode == OPAL_COMMON_UCX_WPMEM_ADDR_EXCHANGE_FULL) {
```
where's the place where you cover the OPAL_COMMON_UCX_WPMEM_ADDR_EXCHANGE_DIRECT case?

janjust replied:
During get_dynamic_win_info, when we evaluate the temp_dynamic_buffer, the mem->addr will be NULL, because it wasn't set during win_attach (with DIRECT). Then we copy over the addresses that we got from the dynamic win.

janjust replied:
This is also very inefficient during rget rput, because we get the dynamic win twice. and we get sevel kb, it's really bad, let's merge the fix, and we can work on a better solution.

janjust commented on ompi/mca/osc/ucx/osc_ucx_passive_target.c:
```diff
@@ -216,20 +216,33 @@ int ompi_osc_ucx_lock_all(int mpi_assert, struct ompi_win_t *win) {
+        }
     }
 
+
```
```suggestion

```

janjust commented on opal/mca/common/ucx/common_ucx_wpool.c:
```diff
@@ -634,11 +637,16 @@ static int _tlocal_ctx_connect(_ctx_record_t *ctx_rec, int target)
         opal_mutex_unlock(&winfo->mutex);
         return OPAL_ERROR;
     }
+
```
```suggestion

```

awlauria committed:
config/pmix: Fix typo bug with internal_pmix_args.

prte -> pmix

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>

awlauria committed:
Enable pmix binaries in ompi builds.

Related to #7285.

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>

awlauria committed:
Merge pull request #9640 from awlauria/make_pmix_apps

Make pmix_info into ompi exports

janjust committed:
Merge pull request #9636 from janjust/master-osc-dbg

master: osc/ucx fixes dynamic windows

ggouaillardet committed:
coll/basic: correctly handle zero size datatypes in scatterv

Since no message is sent/received when the count is zero,
do not send/receive any message when the datatype size is zero.

Refs. open-mpi/ompi#9619

Signed-off-by: Gilles Gouaillardet <gilles@rist.or.jp>

ggouaillardet committed:
coll/basic: correctly handle zero size datatypes in gatherv

Since no message is sent/received when the count is zero,
do not send/receive any message when the datatype size is zero.

Refs. open-mpi/ompi#9619

Signed-off-by: Gilles Gouaillardet <gilles@rist.or.jp>

mkurnosov committed:
Fixes send/recv offsets in rbuf for the case of MPI_IN_PLACE

Signed-off-by: Mikhail Kurnosov <mkurnosov@gmail.com>

awlauria committed:
Merge pull request #9648 from mkurnosov/coll-base-alltoall-fix

coll/base/alltoall: fixes send/recv offsets in rbuf for the case of MPI_IN_PLACE

janjust committed:
Merge pull request #9631 from MamziB/mamzi/oshmem-v1.5-new-apis

OSHMEM v1.5 API header

awlauria committed:
Merge pull request #9644 from ggouaillardet/topic/coll_basic_scatterv

coll/basic: correctly handle zero size datatypes in gatherv/scatterv

awlauria committed:
Merge pull request #9536 from abouteiller/doc/ulfm/components-upgrade

Add a devel document to guide FT/ULFM upgrade of components

janjust committed:
Active target sync

Signed-off-by: Tomislav Janjusic <tomislavj@nvidia.com>
